### PR TITLE
Lessons 4 build and customize platform manager web ui

### PR DESCRIPTION
**Pull request by mvktsk**
Create lesson 4
Fix typo in creating-new-module.md
Remove Script debugin section in lesson 3

Related tasks: CCIT-195

### CODE REVIEW

File: pages/docs/lessons/lesson4.md [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+Use this guide to create Platform Manager UI for the Virto Commerce "Customer Reviews" module (created in the [Lesson 3](/lesson3.md)). There will be a UI created for the list and view product reviews.
+
+After completing this lesson, a Platform Manager UI for the Virto Commerce "Customer Reviews" module will be created, including:
+
+* blade with the customer reviews list for products;
+* widget on the product page to view the number of reviews and open the new reviews blade for the product.
+
+## Video
+
+https://web.microsoftstream.com/video/43fd5a0a-d482-4de9-93af-4e0ad0837601
```
Comment (OlegoO) [inline pages/docs/lessons/lesson4.md]: Upload video to youtube. Microsoft Stream doesn't support public video

File: pages/docs/lessons/lesson4.md [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+* widget on the product page to view the number of reviews and open the new reviews blade for the product.
+
+## Video
+
+https://web.microsoftstream.com/video/43fd5a0a-d482-4de9-93af-4e0ad0837601  
+
+## Prerequisites
+
+* Installed Virto Commerce Platform Manager
+* Visual Studio 2017 or higher
```
Comment (OlegoO) [inline pages/docs/lessons/lesson4.md]: Visual Studio 2019

File: pages/docs/lessons/lesson4.md [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+* Basic JavaScript (Angular) knowledge
+* Passed [Lesson 3](/lesson3.md)
+
+## Glossary
```
Comment (OlegoO) [inline pages/docs/lessons/lesson4.md]: Maybe move to lesson 1 ?

File: pages/docs/lessons/lesson4.md [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+* JS – JavaScript;
+* Platform - Virto Commerce Platform;
+* Platform Manager – Virto Commerce Platform Manager, UI for Platform managing;
+* VC – Virto Commerce;
+* VS – Visual Studio;
+* UI – User Interface.
+
+## Platform Manager UI elements
+
+Platform Manager provides following UI elements:
```
Comment (OlegoO) [inline pages/docs/lessons/lesson4.md]: Add Screenshot and hilight all UI elements

File: pages/docs/lessons/lesson4.md (hunk 1 above)
Comment (mvktsk) [reply pages/docs/lessons/lesson4.md]: Link removed

File: pages/docs/lessons/lesson4.md (hunk 2 above)
Comment (mvktsk) [reply pages/docs/lessons/lesson4.md]: fixed

File: pages/docs/lessons/lesson4.md [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+<div class="gridster-cnt" ng-click="openBlade()">
+    <div class="cnt-inner">
+        <div class="list-count" ng-if="loading">...</div>
+        <div class="list-count" ng-if="!loading">{{ totalCount }}</div>
+        <div class="list-t" ng-if="!loading">{{ 'customerReviews.widgets.item-detail.title' | translate }}</div>
+    </div>
+</div>
+```
+
+4. Add widget to a Product detail blade, register a new widget in *module.js*:
```
Comment (eMazeika) [inline pages/docs/lessons/lesson4.md]: "Inject" the new widget into the Product detail UI, by registering it in *module.js*:

File: pages/docs/lessons/lesson4.md (hunk 3 above)
Comment (mvktsk) [reply pages/docs/lessons/lesson4.md]: I do not think so. Lesson 4 can be used as an independent lesson, and it’s easier to search all in one document than in several.

File: pages/docs/lessons/lesson4.md [hunk 6]
```diff
@@ -0,0 +1,387 @@
...
+
+To debug JS code at run-time use special debugging tools in browser. You can read more about Chrome debug tools and how to debug any JS issue in this [article](https://javascript.info/debugging-chrome). In order to enable JS debugging, change platform's Web.config, app setting **VirtoCommerce:EnableBundlesOptimizations** value to false:
+
+```xml
+<add key="VirtoCommerce:EnableBundlesOptimizations" value="false" />
+```
+
+## Build "Customer reviews" module Web UI
+
+Platform Manager UI elements placed in **\CustomerReviewsModule.Web\Scripts** folder.
```
[the remark this answers is not in the document]
Comment (mvktsk) [reply pages/docs/lessons/lesson4.md]: fixed

File: pages/docs/lessons/lesson4.md [hunk 7]
```diff
@@ -0,0 +1,387 @@
...
+
+Typical structure of this folder is:
+
+* Scripts entry point for Platform Manager user interface;
```
[the remark this answers is not in the document]
Comment (mvktsk) [reply pages/docs/lessons/lesson4.md]: fixed

File: pages/docs/lessons/lesson4.md [hunk 8]
```diff
@@ -0,0 +1,387 @@
...
+  * resources - folder contains definitions for the resources (services) available inside AngularJS module;
+  * blades - folder contains AngularJS controllers and markup for the module blades;
+  * widgets - folder contains AngularJS controllers and markup for the module widgets;
+  * module.js - main entry point for the module UI, containing AngularJS module definition, blades and widgets registrations.
```
[the remark this answers is not in the document]
Comment (mvktsk) [reply pages/docs/lessons/lesson4.md]: fixed

File: pages/docs/lessons/lesson4.md [hunk 9]
```diff
@@ -0,0 +1,387 @@
...
+
+The Platform Manager also allows localizing UI elements. Localization based on resources (translation files) files placed in *\CustomerReviewsModule.Web\Localizations* folder.
```
[the remark this answers is not in the document]
Comment (mvktsk) [reply pages/docs/lessons/lesson4.md]: I dont think that it should be placed to localization-related section.  
This section contain sort description of all Platform Manager UI elements including localization in one place.

fixed

File: pages/docs/lessons/lesson4.md [hunk 10]
```diff
@@ -0,0 +1,387 @@
...
+
+### How to define blade for "Customer reviews" module
```
[the remark this answers is not in the document]
Comment (mvktsk) [reply pages/docs/lessons/lesson4.md]: API placed in separate section **Define API controller for "Customer reviews" module**

fixed

File: pages/docs/lessons/lesson4.md [hunk 11]
```diff
@@ -0,0 +1,387 @@
...
+
+1. To get and update data from backend define *search* and *update* API controller methods in *\CustomerReviewsModule.Web\Scripts\resources\customer-reviews-module-api.js* file:
```
[the remark this answers is not in the document]
Comment (mvktsk) [reply pages/docs/lessons/lesson4.md]: fixed

File: pages/docs/lessons/lesson4.md [hunk 12]
```diff
@@ -0,0 +1,387 @@
...
+angular.module('customerReviewsModule')
+    .factory('customerReviewsModule.webApi', ['$resource', function ($resource) {
+        return $resource('api/CustomerReviewsModule', {}, {
+            search: { method: 'POST', url: 'api/customerReviewsModule/search' },
+            update: { method: 'PUT' }
+        });
+}]);
+
+```
+
```
[the remark this answers is not in the document]
Comment (mvktsk) [reply pages/docs/lessons/lesson4.md]: fixed

File: pages/docs/lessons/lesson4.md [hunk 13]
```diff
@@ -0,0 +1,387 @@
...
+                blade.isLoading = true;
+                reviewsApi.search(blade.getSearchCriteria(), function (data) {
+                    blade.isLoading = false;
+                    $scope.pageSettings.totalItems = data.totalCount;
+                    blade.currentEntities = data.results;
+                });
+            };
+
+
+            blade.headIcon = 'fa-comments';
```
[the remark this answers is not in the document]
Comment (mvktsk) [reply pages/docs/lessons/lesson4.md]: fixed

File: pages/docs/lessons/lesson4.md [hunk 14]
```diff
@@ -0,0 +1,387 @@
...
+        </div>
+    </div>
+</div>
+
+<script type="text/ng-template" id="list.row.html">
+    <div ng-click="grid.appScope.blade.selectNode(row.entity)" ng-repeat="(colRenderIndex, col) in colContainer.renderedColumns track by col.uid" ui-grid-one-bind-id-grid="rowRenderIndex + '-' + col.uid + '-cell'" class="ui-grid-cell" ng-class="{'ui-grid-row-header-cell': col.isRowHeader, '__hovered': grid.appScope.contextMenuEntity === row.entity, '__selected': row.entity.id===grid.appScope.selectedNodeId }" role="{{col.isRowHeader ? 'rowheader' : 'gridcell' }}" ui-grid-cell></div>
+</script>
+```
+
+3. Create a new *reviews-list.js* blade JS file in *\CustomerReviewsModule.Web\Scripts\blades* folder and define *customerReviewsModule.reviewsListController* by adding code to the *reviews-list.js* file:
```
[the remark this answers is not in the document]
Comment (mvktsk) [reply pages/docs/lessons/lesson4.md]: fixed

File: pages/docs/lessons/lesson4.md [hunk 15]
```diff
@@ -0,0 +1,387 @@
...
+                uiGridHelper.initialize($scope, gridOptions, function (gridApi) {
+                    uiGridHelper.bindRefreshOnSortChanged($scope);
+                });
+                bladeUtils.initializePagination($scope.$parent);
+            };
+
+        }]);
+```
+
+4. Register a created controller in the Angular module by editing *\CustomerReviewsModule.Web\Scripts\module.js*:
```
[the remark this answers is not in the document]
Comment (mvktsk) [reply pages/docs/lessons/lesson4.md]: fixed

File: pages/docs/lessons/lesson4.md [hunk 16]
```diff
@@ -0,0 +1,387 @@
...
+                        }
+                    ]
+                });
+        }
+    ])
+
+~~~
+```
+
+5. Define controller permissions by editing *\CustomerReviewsModule.Web\Scripts\module.js*:
```
[the remark this answers is not in the document]
Comment (mvktsk) [reply pages/docs/lessons/lesson4.md]: fixed

File: pages/docs/lessons/lesson4.md [hunk 17]
```diff
@@ -0,0 +1,387 @@
...
+                action: function () { $state.go('workspace.customerReviewsModuleState') },
+                permission: 'CustomerReviewsModule:read'
+            };
+            mainMenuService.addMenuItem(menuItem);
+        }
+    ]);
+~~~
+```
+
+6. Delete *\CustomerReviewsModule.Web\Scripts\blades\hello-world.html* and *\CustomerReviewsModule.Web\Scripts\blades\hello-world.js* files
```
[the remark this answers is not in the document]
Comment (mvktsk) [reply pages/docs/lessons/lesson4.md]: fixed

File: pages/docs/lessons/lesson4.md [hunk 18]
```diff
@@ -0,0 +1,387 @@
...
+        $scope.$watch("blade.itemId", function (id) {
+            filter.productIds = [id];
+
+            if (id) refresh();
+        });
+    }]);
+```
+
+
+3. Create a new *customerReviewWidget.tpl.html* widget markup file in *\CustomerReviewsModule.Web\Scripts\widgets* folder:
```
[the remark this answers is not in the document]
Comment (mvktsk) [reply pages/docs/lessons/lesson4.md]: fixed

File: pages/docs/lessons/lesson4.md (hunk 5 above)
Comment (mvktsk) [reply pages/docs/lessons/lesson4.md]: fixed

File: pages/docs/lessons/lesson4.md [hunk 19]
```diff
@@ -0,0 +1,387 @@
...
+
+## Build "Customer reviews" module Web UI
+
+Platform Manager UI elements placed in **\CustomerReviewsModule.Web\Scripts** folder.
+
+Typical structure of this folder is:
+
+* Scripts entry point for Platform Manager user interface;
+  * resources - folder contains definitions for the resources (services) available inside AngularJS module;
+  * blades - folder contains AngularJS controllers and markup for the module blades;
```
[the remark this answers is not in the document]
Comment (mvktsk) [reply pages/docs/lessons/lesson4.md]: fixed

File: pages/docs/lessons/lesson4.md (hunk 4 above)
Comment (mvktsk) [reply pages/docs/lessons/lesson4.md]: fixed

File: pages/docs/lessons/lesson4.md [hunk 20]
```diff
@@ -0,0 +1,405 @@
...
+description: Lessons 4 - How to build and customize platform manager Web UI (Angular)
+layout: docs
+date: 2020-01-20
+priority: 3
+---
+# How to build and customize Platform Manager Web UI (Angular)
+
+## Summary
+
+Use this guide to create Platform Manager UI for the Virto Commerce "Customer Reviews" module (created in the [Lesson 3](/lesson3.md)). There will be a UI created for the list and view product reviews.
```
Comment (eMazeika) [inline pages/docs/lessons/lesson4.md]: rephrase in some other way "There will be a UI created for the list and view product reviews."

File: pages/docs/lessons/lesson4.md [hunk 21]
```diff
@@ -0,0 +1,405 @@
...
+
+## Build "Customer reviews" module Web UI
+
+Platform Manager UI elements should be placed in **\CustomerReviewsModule.Web\Scripts** folder.
+
+Typical structure of this folder is:
+
+* **Scripts** - the *root* folder for Platform Manager UI related content:
+  * **blades** - folder containing AngularJS controllers and templates for the module blades;
+  * **resources** - folder contains definitions for the resources (services) available inside AngularJS module;
```
Comment (eMazeika) [inline pages/docs/lessons/lesson4.md]: "folder containing AngularJS definitions of the resources (API endpoints) exposed (and used) by this module;"

File: pages/docs/lessons/lesson4.md [hunk 22]
```diff
@@ -0,0 +1,405 @@
...
+  * **widgets** - folder contains AngularJS controllers and markup for the module widgets;
```
Comment (eMazeika) [inline pages/docs/lessons/lesson4.md]: containing

File: pages/docs/lessons/lesson4.md [hunk 23]
```diff
@@ -0,0 +1,405 @@
...
+  * **module.js** -  the main entry point for the module UI, containing AngularJS module definition, registrations for menus, widgets, etc.
```
Comment (eMazeika) [inline pages/docs/lessons/lesson4.md]: redundant space

File: pages/docs/lessons/lesson4.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (eMazeika) [reply pages/docs/lessons/lesson4.md]: "The localization is based on resources (translation files) placed..."

File: pages/docs/lessons/lesson4.md [hunk 24]
```diff
@@ -0,0 +1,405 @@
...
+        "widgets": {
+            "item-detail": {
+                "title": "Customer reviews"
+                }
+            }
+        }
+}
+```
+
+These translations will be used in blade and widget to defining title and fields name in Platform Manager UI.
```
Comment (eMazeika) [inline pages/docs/lessons/lesson4.md]: widget to label titles and field names in Platform Manager UI.

File: pages/docs/lessons/lesson4.md [hunk 25]
```diff
@@ -0,0 +1,405 @@
...
+                });
+        }
+    ])
+
+~~~
+```
+
+When the state is "activated", it "binds" the blade's controller with its template and the new blade is shown.
+
+4. Create a new menu item in main menu for the blade and define menu item permissions by editing *\CustomerReviewsModule.Web\Scripts\module.js*:
```
Comment (eMazeika) [inline pages/docs/lessons/lesson4.md]: but the menu item was added by template already...

File: pages/docs/lessons/lesson4.md [hunk 26]
```diff
@@ -0,0 +1,405 @@
...
+
+Save all changes, restart application and open blade in main menu. You should see existing Customer reviews list.
+
+![Reviews list blade](../../assets/images/docs/reviews-list-blade.png)
+
+### How to define widget for "Customer reviews" module
+
+1. Add new *widgets* folder in *\CustomerReviewsModule.Web\Scripts* folder.
+
+2. Create a new *customerReviewWidget.js* widget JS file in *\CustomerReviewsModule.Web\Scripts\widgets* folder and define *customerReviewsModule.customerReviewWidgetController* by adding code to the *customerReviewWidget.js*:
```
Comment (eMazeika) [inline pages/docs/lessons/lesson4.md]: don't repeat the filename

File: pages/docs/lessons/lesson4.md [hunk 27]
```diff
@@ -0,0 +1,405 @@
...
+    </div>
+</div>
+```
+
+4. "Inject" the new widget into the Product detail UI, by registering it in *module.js*:
+
+```JS
+~~~
+.run(['$rootScope', 'platformWebApp.mainMenuService', 'platformWebApp.widgetService', '$state',
+    function ($rootScope, mainMenuService, widgetService, $state) {
```
Comment (eMazeika) [inline pages/docs/lessons/lesson4.md]: remove unused parameter $rootScope. Should remove from template(s) as well.

File: pages/docs/lessons/lesson4.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (mvktsk) [reply pages/docs/lessons/lesson4.md]: fixed

File: pages/docs/lessons/lesson4.md (hunk 21 above)
Comment (mvktsk) [reply pages/docs/lessons/lesson4.md]: fixed

File: pages/docs/lessons/lesson4.md (hunk 22 above)
Comment (mvktsk) [reply pages/docs/lessons/lesson4.md]: fixed

File: pages/docs/lessons/lesson4.md (hunk 23 above)
Comment (mvktsk) [reply pages/docs/lessons/lesson4.md]: fixed

File: pages/docs/lessons/lesson4.md (hunk 24 above)
Comment (mvktsk) [reply pages/docs/lessons/lesson4.md]: fixed

File: pages/docs/lessons/lesson4.md (hunk 26 above)
Comment (mvktsk) [reply pages/docs/lessons/lesson4.md]: fixed

File: pages/docs/lessons/lesson4.md (hunk 27 above)
Comment (mvktsk) [reply pages/docs/lessons/lesson4.md]: fixed

File: pages/docs/lessons/lesson4.md (hunk 25 above)
Comment (mvktsk) [reply pages/docs/lessons/lesson4.md]: fixed

File: pages/docs/lessons/lesson4.md (hunk 20 above)
Comment (mvktsk) [reply pages/docs/lessons/lesson4.md]: fixed

File: pages/docs/lessons/lesson3.md [hunk 28]
```diff
@@ -68,15 +16,13 @@ After completing this lesson, a new module will be created, including:
 * module API for create, delete, update and search customer reviews;
 * test project for testing module API.
 
-## Video
-
-https://web.microsoftstream.com/video/43fd5a0a-d482-4de9-93af-4e0ad0837601 
+Platform Manager UI for the Virto Commerce "Customer Reviews" module will be created in [Lesson 4](/lesson3.md)
```
[the remark this answers is not in the document]
Comment (mvktsk) [reply pages/docs/lessons/lesson3.md]: fixed

File: pages/docs/lessons/lesson4.md [hunk 29]
```diff
@@ -0,0 +1,405 @@
...
+
+* blade with the customer reviews list for products;
+* widget on the product page to view the number of reviews and open the new reviews blade for the product.
+
+## Prerequisites
+
+* Installed Virto Commerce Platform Manager
+* Visual Studio 2019 or higher
+* Basic JavaScript (AngularJS) knowledge
+* Passed [Lesson 3](/lesson3.md)
```
[the remark this answers is not in the document]
Comment (mvktsk) [reply pages/docs/lessons/lesson4.md]: fixed

File: pages/docs/lessons/lesson4.md [hunk 30]
```diff
@@ -0,0 +1,405 @@
...
+                });
+        }
+    ])
+
+~~~
+```
+
+When the state is "activated", it "binds" the blade's controller with its template and the new blade is shown.
+
+4. Add a new menu item in main menu for the blade and define menu item permissions by editing *\CustomerReviewsModule.Web\Scripts\module.js*:
```
[the remark this answers is not in the document]
Comment (mvktsk) [reply pages/docs/lessons/lesson4.md]: fixed

File: pages/docs/lessons/lesson4.md [hunk 31]
```diff
@@ -0,0 +1,405 @@
...
+
+![Toolbar](../../assets/images/docs/lesson4-blade-toolbar.png)
+
+Read more about toolbar in the [Blade toolbar](https://virtocommerce.com/docs/vc2devguide/working-with-platform-manager/basic-functions/blade-toolbar) article.
+
+### Metaform
+
+A metaform is a placeholder (container) control that renders UI content based on provided metadata.
+
+![Metaform](../../assets/images/docs/lesson4-blade-metaform.png)
```
[the remark this answers is not in the document]
Comment (mvktsk) [reply pages/docs/lessons/lesson4.md]: fixed

File: pages/docs/lessons/lesson4.md [hunk 32]
```diff
@@ -0,0 +1,405 @@
...
+description: Lessons 4 - How to build and customize platform manager Web UI (AngularJS)
+layout: docs
+date: 2020-01-20
+priority: 3
+---
+# How to build and customize Platform Manager Web UI (AngularJS)
+
+## Summary
+
+Use this guide to create Platform Manager UI for the Virto Commerce "Customer Reviews" module (created in the [Lesson 3](/lesson3.md)).
```
[the remark this answers is not in the document]
Comment (mvktsk) [reply pages/docs/lessons/lesson4.md]: fixed

File: pages/docs/lessons/lesson4.md (hunk 31 above)
[the remark this answers is not in the document]
Comment (eMazeika) [reply pages/docs/lessons/lesson4.md]: Let's be honest: this UI couldn't be build using metaform. Can check https://github.com/VirtoCommerce/vc-module-catalog/blob/dev/VirtoCommerce.CatalogModule.Web/Scripts/blades/item-detail.tpl.html
Let's use another example, (from metaform article maybe).

[the remark this answers is not in the document]
Comment (mvktsk) [reply pages/docs/lessons/lesson4.md]: fixed

**Pull request merged**